### PR TITLE
Adds Cards Against Humanity/Spess to Library

### PR DESCRIPTION
**Pull request by Crystalwarrior**
Based on https://github.com/tgstation/-tg-station/pull/15872 this is a PR containing SS13-themed Cards Against Humanity.
A box with 4 decks is added to Library:
1 white deck
1 black deck
1 white blanks deck
1 black blanks deck

Blanks can be filled with a pen for maximum customizability and fun.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,13 +2,13 @@
 
 #Ignore everything in datafolder and subdirectories
 /data/**/*
+/config/**/*
```
Comment (Carbonhell) [inline .gitignore]: won't this disallow us from editing initial config files

Comment (Crystalwarrior) [inline .gitignore]: Oh yeah, I can fix that.

File: code/game/objects/items/toys.dm [hunk 2]
```diff
@@ -830,12 +825,7 @@
...
 			user << "<span class='warning'>You can't mix cards from other decks!</span>"
-		if(cards.len > 26)
-			src.icon_state = "deck_[deckstyle]_full"
-		else if(cards.len > 10)
-			src.icon_state = "deck_[deckstyle]_half"
-		else if(cards.len > 1)
-			src.icon_state = "deck_[deckstyle]_low"
+	update_icon()
 
 /obj/item/toy/cards/deck/MouseDrop(atom/over_object)
```
Comment (Carbonhell) [inline code/game/objects/items/toys.dm:830]: unrelated to this pr but if i remember correctly there's no way to make a deck out of normal cards and i think it's ridicolous,any chance you can fix that?

File: code/modules/games/cas.dm [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+// Original code by Zuhayr, Polaris Station, ported with modifications
+
+/obj/item/toy/cards/deck/cas
+	name = "\improper CAS deck (white)"
+	desc = "A deck for the game Cards Against Spess, still popular after all these centuries. Warning: may include traces of broken fourth wall.\nThis is the white deck."
+	icon = 'icons/obj/toy.dmi'
+	icon_state = "deck_caswhite_full"
+	deckstyle = "caswhite"
+	var/blanks = 0
+	var/decksize = 100
```
Comment (Carbonhell) [inline code/modules/games/cas.dm:14]: feels like a var that should be part of all cards

File: code/game/objects/items/toys.dm (hunk 2 above)
Comment (Crystalwarrior) [inline code/game/objects/items/toys.dm:830]: I think it's better to be done in a different PR

File: code/modules/games/cas.dm (hunk 3 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:14]: decks have unlimited size, this is just for generating cards it contains initially

File: code/modules/games/cas.dm [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+/obj/item/toy/cards/deck/cas/black/blanks
+	name = "\improper CAS deck (black blanks)"
+	desc = "A deck for the game Cards Against Spess, still popular after all these centuries. Warning: may include traces of broken fourth wall.\nThis is a black deck which contains blanks to be used in custom games of CAS."
+	decksize = 0
+	blanks = 50
+
+/obj/item/toy/cards/deck/cas/New()
+	allcards = file2list(card_text_file)
+	var/list/possiblecards = allcards.Copy()
```
Comment (Carbonhell) [inline code/modules/games/cas.dm:41]: wouldn't it be better to shuffle the list so you will get different cards?

File: code/modules/games/cas.dm [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	if(possiblecards.len < decksize) // sanity check
+		decksize = (possiblecards.len - 1)
+	var/list/randomcards = list()
+	while (randomcards.len < decksize)
+		randomcards += pick_n_take(possiblecards)
+	for(var/i=1 to randomcards.len)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm]: for(var/i in 1 to randomcards.len), it's faster process wise,just a minor thing tho

Comment (Crystalwarrior) [inline code/modules/games/cas.dm]: are you telling me to remove that odd while loop?

File: code/modules/games/cas.dm [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		var/cardtext = randomcards[i]
+		cards += "[cardtext]"
+	if(!blanks)
+		return
+	for(var/x=1 to blanks)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm]: same of up here

File: code/modules/games/cas.dm [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		cards += "Blank Card"
+
+/obj/item/toy/cards/deck/cas/attack_hand(mob/user)
+	if(user.lying)
+		return
+	var/choice = null
+	if(cards.len == 0)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm]: if(!cards.len)

File: code/modules/games/cas.dm [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	for(var/i=1 to randomcards.len)
+		var/cardtext = randomcards[i]
+		cards += "[cardtext]"
+	if(!blanks)
+		return
+	for(var/x=1 to blanks)
+		cards += "Blank Card"
+
+/obj/item/toy/cards/deck/cas/attack_hand(mob/user)
+	if(user.lying)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm:56]: isn't this check redundant?huh

File: code/modules/games/cas.dm [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	choice = cards[1]
+	H.name = "[copytext(deckstyle,4)] [H.name]"
+	if(deckstyle == "caswhite")
+		H.pixel_x = 5
+	H.cardname = choice
+	if(choice == "Blank Card")
+		H.blank = 1
+	H.parentdeck = src
+	H.icon_state = "singlecard_[deckstyle]"
+	src.cards -= choice
```
Comment (Carbonhell) [inline code/modules/games/cas.dm]: src. SRC. SSSSRRRRRCCCCCCC...........

File: code/modules/games/cas.dm [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+	H.pickup(user)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm:73]: pickup?isn't it put_in_hand or something similar?

File: code/modules/games/cas.dm (hunk 8 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:56]: ...what? no it isn't

File: code/modules/games/cas.dm [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+			user << "<span class='notice'>The card reads: [src.cardname]</span>"
+		else
+			user << "<span class='warning'>You need to get closer to check the card!</span>"
+
+/obj/item/toy/cards/singlecard/cas/Flip()
+	return
+
+/obj/item/toy/cards/singlecard/cas/attackby(obj/item/I, mob/living/user, params)
+	if(istype(I, /obj/item/weapon/pen))
+		if(!blank)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm:102]: this blank var is a bit pointless,you could just use istype(I, whatevertheblankcardpathis)

File: code/modules/games/cas.dm [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+			user << "You cannot write on that card."
+			return
+		var/cardtext = sanitize(input(user, "What do you wish to write on the card?", "Card Writing") as text|null, 50)
```
Comment (Carbonhell) [inline code/modules/games/cas.dm]: sanitize(input) ? isn't stripped_input better

File: code/modules/games/cas.dm (hunk 4 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:41]: you can do that if you activate the deck in-hand.

File: code/game/objects/items/toys.dm (hunk 2 above)
Comment (Carbonhell) [inline code/game/objects/items/toys.dm:830]: alright

File: code/modules/games/cas.dm (hunk 3 above)
Comment (Carbonhell) [inline code/modules/games/cas.dm:14]: oh k

File: code/modules/games/cas.dm (hunk 4 above)
Comment (Carbonhell) [inline code/modules/games/cas.dm:41]: fair enough

File: code/modules/games/cas.dm (hunk 8 above)
Comment (Carbonhell) [inline code/modules/games/cas.dm:56]: i though attack_hand checked for user.stat, user.lying, stuns and weakens,etc already.

File: code/modules/games/cas.dm (hunk 10 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:73]: This is copy-pasted from toys.dm decks, soooo..

File: strings/cas_black.txt [hunk 13]
```diff
@@ -0,0 +1,86 @@
+Today, Security killed ____.
+Security, the clown's breaking into ____.
+The Chaplain is worshiping the Church of _____.
```
Comment (Carbonhell) [inline strings/cas_black.txt]: worshipping*

File: strings/cas_black.txt [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+What angered the gods this time?
+Why didn't engineering set up the singularity?
+What does the traitor steal this time?
```
Comment (Carbonhell) [inline strings/cas_black.txt]: more like What will the traitor steal this time?

File: code/modules/games/cas.dm (hunk 11 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:102]: uhhh no, because there's no "blank" single card type. It's only for decks.

File: code/modules/games/cas.dm (hunk 12 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm]: Gonna use that

File: strings/cas_black.txt (hunk 13 above)
Comment (Crystalwarrior) [inline strings/cas_black.txt]: lel gg

File: strings/cas_black.txt [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+New in syndicate uplinks: __________!
+NanoTrasen's labor union decided to use _______ to raise employee morale.
+The Chemist's drug of choice is ______
+According to space law, ________ is now punishable by ______.
+________ is why I'm afraid of maintenance.
+________ is the most robust person on the station.
+What made the nuclear operatives fluke?
+Every spaceman receives an internals box and ________.
+What brought the space orgy to a grinding halt?
+Scientists are no longer allowed to make  _____.
```
Comment (Carbonhell) [inline strings/cas_black.txt]: double space after "make"

File: strings/cas_black.txt [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+What's the worst thing an abductor can give you?
+One day while Andy was __________, Woody got _________.
+A good engineer always carries ________.
+Why is the station so filthy?
+THE SHUTTLE CANNOT BE CALLED BECAUSE OF ________ ON THE STATION.
+After murdering dozens, the traitor was finally brought down by _________.
+The librarian has read a lot of porn, but never about _____________!
+Oh god! There's ________________ outside escape!
+The hottest new space couple is ________ and __________.
+Lord Singulo hungers for _______.
```
Comment (Carbonhell) [inline strings/cas_black.txt]: Lord Singuloth

**Comment by harmonyn** [comment]
Yo, so, how many of these changes should I implement in my source PR?

**Comment by Crystalwarrior** [comment]
all of them :^)

File: code/game/objects/items/toys.dm [hunk 17]
```diff
@@ -770,26 +770,29 @@
...
+/obj/item/toy/cards/deck/update_icon()
 	if(cards.len > 26)
-		src.icon_state = "deck_[deckstyle]_full"
+		icon_state = "deck_[deckstyle]_full"
 	else if(cards.len > 10)
-		src.icon_state = "deck_[deckstyle]_half"
-	else if(cards.len > 1)
-		src.icon_state = "deck_[deckstyle]_low"
+		icon_state = "deck_[deckstyle]_half"
+	else if(cards.len > 0)
```
Comment (Carbonhell) [inline code/game/objects/items/toys.dm:792]: else if(cards.len)

File: code/game/objects/items/toys.dm [hunk 18]
```diff
@@ -770,26 +770,29 @@
...
+		icon_state = "deck_[deckstyle]_low"
+	else if(!cards.len)
```
Comment (Carbonhell) [inline code/game/objects/items/toys.dm:794]: else

File: code/modules/games/cas.dm (hunk 11 above)
Comment (Carbonhell) [inline code/modules/games/cas.dm:102]: ah alright

**Comment by Carbonhell** [comment]
code seems acceptable

File: code/modules/games/cas.dm (hunk 4 above)
Comment (harmonyn) [inline code/modules/games/cas.dm:41]: also pick_n_take effectively randomizes the deck

**Comment by calibraptor** [comment]
Seems neat

**Comment by Carbonhell** [comment]
did you fix all the bugs i said,if yes code's ok

**Comment by Crystalwarrior** [comment]
Yeah I did, check commits

**Comment by Lawligagger** [comment]
yes please

**Comment by doobiedoo23** [comment]
Crystal coded something that isn't an overhaul? :+1: :+1: :+1:

**Comment by Crystalwarrior** [comment]
nerd

**Comment by Nokington** [comment]
Sure

**Comment by Crystalwarrior** [comment]
This is merge-ready

File: code/game/objects/items/toys.dm [hunk 19]
```diff
@@ -770,26 +770,29 @@
...
 	choice = cards[1]
 	H.cardname = choice
 	H.parentdeck = src
-	H.add_fingerprint(user)
 	var/O = src
 	H.apply_card_vars(H,O)
 	src.cards -= choice
 	H.pickup(user)
-	user.put_in_active_hand(H)
+	user.put_in_hands(H)
```
Comment (JohnGinnane) [inline code/game/objects/items/toys.dm:783]: I'm on mobile, but what is the difference here?

Comment (Crystalwarrior) [inline code/game/objects/items/toys.dm:783]: put_in_active_hand drops the item if active hand is full, when put_in_hands check if the other hand is free first.

File: code/modules/games/cas.dm [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+	blanks = 50
+
+/obj/item/toy/cards/deck/cas/New()
+	allcards = file2list(card_text_file)
+	var/list/possiblecards = allcards.Copy()
+	if(possiblecards.len < decksize) // sanity check
+		decksize = (possiblecards.len - 1)
+	var/list/randomcards = list()
+	while (randomcards.len < decksize)
+		randomcards += pick_n_take(possiblecards)
```
Comment (JohnGinnane) [inline code/modules/games/cas.dm:46]: This could lead to potentially duplicates, should we use |= instead?

File: code/modules/games/cas.dm [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+		H.blank = 1
+	H.parentdeck = src
+	H.icon_state = "singlecard_[deckstyle]"
+	cards -= choice
+	H.pickup(user)
+	user.put_in_hands(H)
+	user.visible_message("[user] draws a card from the deck.", "<span class='notice'>You draw a card from the deck.</span>")
+	update_icon()
+
+/obj/item/toy/cards/deck/cas/update_icon()
```
Comment (JohnGinnane) [inline code/modules/games/cas.dm:78]: What's wrong with using parent proc, the one defined above?

File: code/modules/games/cas.dm (hunk 20 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:46]: Ehh, I'm pretty sure it's fine as is

File: code/game/objects/items/toys.dm (hunk 19 above)
Comment (JohnGinnane) [inline code/game/objects/items/toys.dm:783]: Ah very clever

File: code/modules/games/cas.dm (hunk 21 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:78]: Lacks the blank card stuffs

File: code/modules/games/cas.dm [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+		icon_state = "deck_[deckstyle]_empty"
+
+/obj/item/toy/cards/singlecard/cas
+	name = "CAS card"
+	desc = "A CAS card."
+	icon_state = "cas_white"
+	var/blank = 0
+
+/obj/item/toy/cards/singlecard/cas/examine(mob/user)
+	if(ishuman(user))
```
Comment (JohnGinnane) [inline code/modules/games/cas.dm:91]: Why can't monkeys read? :(

Comment (JohnGinnane) [inline code/modules/games/cas.dm:91]: Or cyborgs

Comment (Crystalwarrior) [inline code/modules/games/cas.dm:91]: blame original card code

Comment (JohnGinnane) [inline code/modules/games/cas.dm:91]: Does this mean you'll change it?

**Comment by JohnGinnane** [comment]
Code looks good, just please allow anyone to read it, not just humans I think

File: code/modules/games/cas.dm (hunk 22 above)
Comment (Crystalwarrior) [inline code/modules/games/cas.dm:91]: nop, I'm working on other stuffs tbh

**Comment by Crystalwarrior** [comment]
Codecheck + Approval received, merging...
